module/Photo/src/Photo/Service/AlbumCover.php

```diff
@@ -0,0 +1,158 @@
+        $newPath = $this->getPhotoService()->generateStoragePath($tempFileName);
+        $config = $this->getConfig();
+        rename($tempFileName, $config['upload_dir'] . '/' . $newPath);
+        return $newPath;
+    }
+
+    protected function generateCover($album)
+    {
+        $config = $this->getConfig();
+        $columns = $config['album_cover']['cols'];
+        $rows = $config['album_cover']['rows'];
+        $count = $columns * $rows;
+        $images = $this->getImages($album, $count);
+        //if there are not enough images to fill the matrix, reduce the rows and columns
+        while (($columns > 1 || $rows > 1) && count($images) < $count) {
+            if ($columns < $rows) {
+                $rows--;
+            } else {
+                $columns--;
+            }
+            $count = $rows * $columns;
+        }
+        $innerWidth = $config['album_cover']['width'] - 2 * $config['album_cover']['outer_border'];
+        $innerHeight = $config['album_cover']['height'] - 2 * $config['album_cover']['outer_border'];
+        $imageWidth = floor(($innerWidth - ($columns - 1) * $config['album_cover']['inner_border']) / $columns);
+        $imageHeight = floor(($innerHeight - ($rows - 1) * $config['album_cover']['inner_border']) / $rows);
+        //increase outer border due to flooring of image dimensions
+        $outerBorderX = $config['album_cover']['outer_border'] + ceil(($innerWidth - ($columns * $imageWidth + ($columns - 1) * $config['album_cover']['inner_border'])) / 2);
+        $outerBorderY = $config['album_cover']['outer_border'] + ceil(($innerHeight - ($rows * $imageHeight + ($rows - 1) * $config['album_cover']['inner_border'])) / 2);
+        // Make a blank canvas
+        $target = new Imagick();
+        $target->newImage($config['album_cover']['width'], $config['album_cover']['height'], $config['album_cover']['background']);
+        $index = 0;
+        if (count($images) >= $count) {
+            for ($x = 0; $x < $columns; $x++) {
+                for ($y = 0; $y < $rows; $y++) {
+                    $image = $this->resizeCropImage($images[$index], $imageWidth, $imageHeight);
+                    $index++;
+                    $target->compositeImage($image, imagick::COMPOSITE_COPY, ($imageWidth + $config['album_cover']['inner_border']) * $x + $outerBorderX, ($imageHeight + $config['album_cover']['inner_border']) * $y + $outerBorderY);
+                }
+            }
+        }
+        $target->setImageFormat("png");
+        return $target;
+    }
+
+    /**
+     * Specialized function to rezie and crop photos such that they always
+     * fill the full width and height without damaging the aspect ratio of the
+     * photo.
+     * 
+     * @param Imagick $image The Imagick object to be resized and cropped
+     * @param int $width The desired width
+     * @param int $height The desired height
+     * @return Imagick $image
+     */
+    protected function resizeCropImage($image, $width, $height)
```

Comment:
Helaas kan dit niet, omdat Imagick::thumbnailimage geen mooie resultaten oplevert voor foto's met een afwijkende aspect ratio. Als de aspect ratio afwijkt worden bij thumbnailimage of de aspect ratio veranderd of de image krijgt niet de aangegeven breedte en hoogte. Bij deze resizeCropImage functie wordt dat probleem opgelost. Deze functie kan waarschijnlijk wel korter worden door thumbnailimage te gebruiken voor images die wel een normale aspect ratio hebben.